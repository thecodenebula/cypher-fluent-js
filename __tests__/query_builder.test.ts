import {
  call,
  callQ,
  contains,
  endsWith,
  eq,
  exists,
  gt,
  gte,
  inList,
  isNull,
  lt,
  lte,
  match,
  node,
  nodeLabel,
  not,
  pattern,
  raw,
  regex,
  relationship,
  startsWith,
  value,
  variable,
} from '../src';

describe('query builder', () => {
  it('should match basic queries', () => {
    const q1 = match(node('actor', 'Actor'), node('person', 'Person')).return('actor', 'person');
    expect(q1.build()).toMatchSnapshot();

    const q2 = match(node('n1')).optionalMatch(node('n2')).return('n1');
    expect(q2.build()).toMatchSnapshot();

    const q3 = match(relationship(node('n1'), node('n2'))).return('n1');
    expect(q3.build()).toMatchSnapshot();

    const q4 = match(relationship(relationship(node('n1'), node('n2')), node('n3'))).return('n1');
    expect(q4.build()).toMatchSnapshot();

    const q5 = match(relationship(node('n1'), node('n2')).lr()).return('n1');
    expect(q5.build()).toMatchSnapshot();

    const q6 = match(relationship(node('n1'), node('n2')).withType('LOVES')).return('n1');
    expect(q6.build()).toMatchSnapshot();

    const q7 = match(node('actor', 'Actor'), node('person', 'Person'))
      .return('actor', 'person')
      .skip(10)
      .limit(100)
      .orderBy('actor.name asc');
    expect(q7.build()).toMatchSnapshot();

    expect(q7.skip(variable('skip')).limit(variable('limit')).build()).toMatchSnapshot();
  });

  it('should do basic where statements', () => {
    const q1 = match(node('n'))
      .where(eq(raw('n.name'), 'Mike'))
      .return('n');
    expect(q1.build()).toMatchSnapshot();

    const q2 = match(node('n'))
      .where(eq(raw('n.name'), variable('name')))
      .return('n');
    expect(q2.build()).toMatchSnapshot();

    const composedWhere = eq(raw('n.name'), variable('name')).and(eq(raw('n.age'), 30));
    const otherWhere = eq(raw('n.city'), variable('city')).or(eq(raw('n.country'), variable('country')));

    const q3 = match(node('n')).where(composedWhere).where(otherWhere).return('n');
    expect(q3.build()).toMatchSnapshot();

    const q4 = match(node('n')).where(composedWhere.or(otherWhere)).return('n');
    expect(q4.build()).toMatchSnapshot();

    const q5 = match(node('m'))
      .where(
        isNull(raw('m.name')),
        startsWith(raw('m.name'), 'rafa'),
        endsWith(raw('m.name'), 'rafa'),
        contains(raw('m.name'), 'rafa'),
        regex(raw('m.name'), 'Tim.*'),
        eq(raw('m.name'), variable('name')),
        inList(raw('m.name'), variable('name')),
        gt(raw('m.age'), variable('age')),
        gte(raw('m.age'), variable('age')),
        lt(raw('m.age'), variable('age')),
        lte(raw('m.age'), variable('age'))
      )
      .return('m');
    expect(q5.build(true)).toMatchSnapshot();

    const q6 = match(node('m'))
      .where(not(eq(raw('m.name'), 'John')))
      .return('m');
    expect(q6.build()).toMatchSnapshot();

    const q7 = match(node('m')).where(nodeLabel('m', 'Actor')).return('m');
    expect(q7.build()).toMatchSnapshot();

    const q8 = match(node('m')).where(nodeLabel('m', 'Actor', 'Person')).return('m');
    expect(q8.build()).toMatchSnapshot();

    const q9 = match(node('m', 'Person'))
      .where(exists(relationship(node('m'), node(), '', 'LOVES')))
      .return('m');
    expect(q9.build()).toMatchSnapshot();

    const q10 = match(node('m')).where(exists(match(relationship(node('m'), node(), '', 'LOVES'))));
    expect(q10.return('m').build()).toMatchSnapshot();

    const q11 = match(node('m'))
      .where(pattern(relationship(node('m'), node(), '', 'LOVES')))
      .return('m');
    expect(q11.build()).toMatchSnapshot();

    const q12 = match(node('m')).where(exists(q10)).return('m');
    expect(q12.build()).toMatchSnapshot();
  });

  it('should render call queries', () => {
    const builtQuery = callQ(
      call('db.index.fulltext.queryNodes').args(value('fullText'), variable('q')).yield('node', 'score')
    )
      .where(inList(variable('labelRestriction'), raw('labels(node)')))
      .return('node', 'labels(node)', 'score')
      .orderBy('score DESC')
      .skip(variable('offset'))
      .limit(variable('limit'));
    expect(builtQuery.build()).toMatchSnapshot();
  });
});
