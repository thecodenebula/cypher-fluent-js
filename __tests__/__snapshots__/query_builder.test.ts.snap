// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query builder should do basic where statements 1`] = `
"MATCH (n)
WHERE ((n.name = 'Mike'))
RETURN n"
`;

exports[`query builder should do basic where statements 2`] = `
"MATCH (n)
WHERE ((n.name = $name))
RETURN n"
`;

exports[`query builder should do basic where statements 3`] = `
"MATCH (n)
WHERE ((n.name = $name) AND ((n.age = 30))) AND (((n.city = $city) OR ((n.country = $country))))
RETURN n"
`;

exports[`query builder should do basic where statements 4`] = `
"MATCH (n)
WHERE ((n.name = $name) AND ((n.age = 30)) OR ((n.city = $city) OR ((n.country = $country))))
RETURN n"
`;

exports[`query builder should do basic where statements 5`] = `
"MATCH (m)
WHERE ((m.name IS NULL))
AND
((m.name starts with 'rafa'))
AND
((m.name ends with 'rafa'))
AND
((m.name contains 'rafa'))
AND
((m.name =~ 'Tim.*'))
AND
((m.name = $name))
AND
((m.name in $name))
AND
((m.age > $age))
AND
((m.age >= $age))
AND
((m.age < $age))
AND
((m.age <= $age))
RETURN m"
`;

exports[`query builder should do basic where statements 6`] = `
"MATCH (m)
WHERE ((NOT (m.name = 'John')))
RETURN m"
`;

exports[`query builder should do basic where statements 7`] = `
"MATCH (m)
WHERE ((m:Actor))
RETURN m"
`;

exports[`query builder should do basic where statements 8`] = `
"MATCH (m)
WHERE ((m:Actor:Person))
RETURN m"
`;

exports[`query builder should do basic where statements 9`] = `
"MATCH (m:Person)
WHERE ((exists((m)-[:LOVES]-())))
RETURN m"
`;

exports[`query builder should do basic where statements 10`] = `
"MATCH (m)
WHERE ((exists{ MATCH (m)-[:LOVES]-() }))
RETURN m"
`;

exports[`query builder should do basic where statements 11`] = `
"MATCH (m)
WHERE (((m)-[:LOVES]-()))
RETURN m"
`;

exports[`query builder should do basic where statements 12`] = `
"MATCH (m)
WHERE ((exists{ MATCH (m)
WHERE ((exists{ MATCH (m)-[:LOVES]-() })) }))
RETURN m"
`;

exports[`query builder should match basic queries 1`] = `
"MATCH (actor:Actor), (person:Person)
RETURN actor, person"
`;

exports[`query builder should match basic queries 2`] = `
"MATCH (n1)
OPTIONAL MATCH (n2)
RETURN n1"
`;

exports[`query builder should match basic queries 3`] = `
"MATCH (n1)--(n2)
RETURN n1"
`;

exports[`query builder should match basic queries 4`] = `
"MATCH (n1)--(n2)--(n3)
RETURN n1"
`;

exports[`query builder should match basic queries 5`] = `
"MATCH (n1)-->(n2)
RETURN n1"
`;

exports[`query builder should match basic queries 6`] = `
"MATCH (n1)-[:LOVES]-(n2)
RETURN n1"
`;

exports[`query builder should match basic queries 7`] = `
"MATCH (actor:Actor), (person:Person)
RETURN actor, person
ORDER BY actor.name asc
SKIP 10
LIMIT 100"
`;

exports[`query builder should match basic queries 8`] = `
"MATCH (actor:Actor), (person:Person)
RETURN actor, person
ORDER BY actor.name asc
SKIP $skip
LIMIT $limit"
`;

exports[`query builder should render call queries 1`] = `
"CALL db.index.fulltext.queryNodes YIELD node, score
WHERE (($labelRestriction in labels(node)))
RETURN node, labels(node), score
ORDER BY score DESC
SKIP $offset
LIMIT $limit"
`;
